src/qibo/models/dbi/utils_scheduling.py

```diff
@@ -0,0 +1,260 @@
+    # coefficients for rotation with [W,H] and H
+    c1 = exp_list.reshape((-1, 1, 1)) * sigma_Gamma_list[1:]
+    c2 = exp_list.reshape((-1, 1, 1)) * sigma_Gamma_list[:-1]
+    # product coefficient
+    trace_coefficients = [0] * (2 * n + 1)
+    for k in range(n + 1):
+        for j in range(n + 1):
+            power = k + j
+            product_matrix = c1[k] @ c2[j]
+            trace_coefficients[power] += 2 * np.trace(product_matrix)
+    # coefficients from high to low (n:0)
+    coef = list(reversed(trace_coefficients[: n + 1]))
+    return coef
+
+def least_squares_polynomial_expansion_coef(dbi_object, d, n):
+    if d is None:
+        d = dbi_object.diagonal_h_matrix
+    # generate Gamma's where $\Gamma_{k+1}=[W, \Gamma_{k}], $\Gamma_0=H
+    Gamma_list = dbi_object.generate_Gamma_list(n+1, d)
+    exp_list = np.array([1 / math.factorial(k) for k in range(n + 1)])
+    # coefficients
+    coef = np.empty(n)
+    for i in range(n):
+        coef[i] = np.real(exp_list[i]*np.trace(d@Gamma_list[i+1]))
+    coef = list(reversed(coef))
+    return coef
+
+#TODO: add a general expansion formula not stopping at 3rd order
+def energy_fluctuation_polynomial_expansion_coef(dbi_object, d, n, state):
+    if d is None:
+        d = dbi_object.diagonal_h_matrix
+    # generate Gamma's where $\Gamma_{k+1}=[W, \Gamma_{k}], $\Gamma_0=H
+    Gamma_list = dbi_object.generate_Gamma_list(n+1, d)
+    # coefficients
+    coef = np.empty(3)
+    coef[0] = np.real(2*covariance(Gamma_list[0], Gamma_list[1],state))
+    coef[1] = np.real(2*variance(Gamma_list[1],state)+2*covariance(Gamma_list[0],Gamma_list[2],state))
+    coef[2] = np.real(covariance(Gamma_list[0], Gamma_list[3],state)+3*covariance(Gamma_list[1], Gamma_list[2],state))
+    coef = list(reversed(coef))
+    return coef
+
+def dGamma_diDiagonal(dbi_object, d, H, n, i,dGamma, Gamma_list):
+    # Derivative of gamma with respect to diagonal elements of D (full-diagonal ansatz)
+    A = np.zeros(d.shape)
+    A[i,i] = 1
+    B = commutator(commutator(A,H),Gamma_list[n-1])
+    W = commutator(d,H)
+    return B + commutator(W,dGamma[-1])
+
+def dpolynomial_diDiagonal(dbi_object, d,H,i):
+    # Derivative of polynomial approximation of potential function with respect to diagonal elements of d (full-diagonal ansatz)
+    # Formula can be expanded easily to any order, with n=3 corresponding to cubic approximation
+    derivative = 0
+    s = polynomial_step(dbi_object, n=3, d=d)
+    A = np.zeros(d.shape)
+    Gamma_list = dbi_object.generate_Gamma_list(4, d)
+    A[i,i] = 1
+    dGamma = [commutator(A,H)]
+    derivative += np.real(np.trace(Gamma_list[0]@A)+np.trace(dGamma[0]@d+Gamma_list[1]@A)*s)
+    for n in range(2,4):
+        dGamma.append(dGamma_diDiagonal(dbi_object,d,H,n,i,dGamma,Gamma_list))
+        derivative += np.real(np.trace(dGamma[-1]@d + Gamma_list[n]@A)*s**n/math.factorial(n))
+
+    return derivative
+
+def gradientDiagonal(dbi_object,d,H):
```

Comment:
```suggestion
def gradientDiagonalEntries(dbi_object,d,H):
```

would be better to have more descriptive names